lib/dry/types/map.rb

```diff
@@ -0,0 +1,110 @@
+module Dry
+  module Types
+    class Map < Definition
+
+      def initialize(keys: Undefined, values: Undefined, meta: EMPTY_HASH)
+        super(::Hash, keys: keys, values: values, meta: meta)
+        validate_options!
+      end
+
+      # @return [Type, Undefined]
+      def keys
+        options[:keys]
+      end
+
+      # @return [Type, Undefined]
+      def values
+        options[:values]
+      end
+
+      # @return [String]
+      def name
+        "Map"
+      end
+
+      # @param [Hash] hash
+      # @return [Hash]
+      def call(hash)
+        try(hash) do |failure|
+          raise MapError, failure.error.join("\n")
+        end.input
+      end
+      alias_method :[], :call
+
+      # @param [Hash] hash
+      # @return [Boolean]
+      def valid?(hash)
+        coerce(hash).success?
+      end
+      alias_method :===, :valid?
+
+      # @param [Hash] hash
+      # @return [Result]
+      def try(hash)
+        result = coerce(hash)
+        return result if result.success? || !block_given?
+        yield(result)
+      end
+
+      # @param meta [Boolean] Whether to dump the meta to the AST
+      # @return [Array] An AST representation
+      def to_ast(meta: true)
+        set = options.reject{ |_,v| Undefined.eql?(v) }
+        set.keys.each do |key|
+          set[key] = set[key].to_ast(meta: meta)
+        end
+        [:map, [set, meta ? self.meta : EMPTY_HASH]]
+      end
+
+      # @param [Hash] new_options
+      # @return [Map]
+      def with(new_options)
+        self.class.new(**options, meta: @meta, **new_options)
+      end
+
+      private
+
+      def if_undefined(receiver)
+        Undefined.eql?(receiver) ? yield : receiver
+      end
+
+      def coerce(input)
+        return failure(
+          input, "#{input.inspect} must be an instance of #{primitive}"
+        ) unless primitive?(input)
+
+        key_type = if_undefined(keys){ Any  }
```

Comment:
With `Any` as a default these two lines shouldn't be needed.

And JFYI I added `Undefined.default` recently: https://github.com/dry-rb/dry-core/blob/master/CHANGELOG.md#added